# Finance with Python video 2 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


## Option Replication and Pricing

In [2]:
S = np.array((20., 5.))

In [3]:
S0 = 10.

In [4]:
B = np.array((11.,11.))

In [5]:
B0 = 10.

In [6]:
M = np.array((S, B)).T
M

array([[20., 11.],
       [ 5., 11.]])

In [7]:
M0 = np.array((S0,B0)).T
M0

array([10., 10.])

In [8]:
0.5 * S - 0.5 * B

array([ 4.5, -3. ])

In [9]:
K = 15 # strike price of call option
C = np.maximum (S - K, 0) # payoff of call option
C

array([5., 0.])

In [10]:
phi = np.linalg.solve(M, C) # phi = solve for the optimal portfolio
phi

array([ 0.33333333, -0.15151515])

above shows we should long .33 bond and short the bond .1515
this is not replicable in the real world because of partial shares isssues

In [11]:
np.dot(M, phi).round(3) # portfolio payoff = option payoff

array([5., 0.])

In [12]:
C0 = np.dot(M0, phi) # cost to set up portfolio
C0

1.8181818181818188

1.818 is the arbitrage free price of the option 

## Fundamental Theorem of Asset Pricing

Is a martingale measure and the absence of arbitrage are equivalent
http://pi.math.cornell.edu/~mec/Summer2008/spulido/fftap.html

### Probability Measure and Expectation 

In [13]:
P = 0.5
P = np.array((P, 1-P))
P

array([0.5, 0.5])

In [14]:
ES = np.dot(S, P)
ES

12.5

In [15]:
EB = np.dot(B,P)
EB

11.0

### Maritingale Property 

In [16]:
i = 0.1

In [17]:
df = 1 / (1+i)
df

0.9090909090909091

Below used $ for LaTeX. Must use markdown for LaTeX

$Q$ is called a Martingale measure if 

$\frac{1}{1+i}\cdot E^Q(S_1) = S_0$

holds

In [18]:
df * np.dot(S, P)

11.363636363636363

In [19]:
S0 # we are looking for S0, the above did not give 10.0
# under P, the price process was not a price process

10.0

## Deriving a Martingale Measure 

In [20]:
from scipy.optimize import minimize

In [21]:
def error (Q):
    return (np.dot(S, Q) / (1+i) - S0) ** 2

In [22]:
error (P)

1.859504132231404

In [23]:
# constraint
cons = ({'type': 'eq', 'fun': lambda Q: Q.sum() - 1}) #this is a probability constraint

In [24]:
# boundaries
bnds = ((0,1),(0,1))

In [25]:
res = minimize(error,(0.5, 0.5), bounds=bnds, constraints=cons)
res

     fun: 1.0172028584007344e-14
     jac: array([ 1.25849558e-06, -6.09001821e-07])
 message: 'Optimization terminated successfully.'
    nfev: 9
     nit: 2
    njev: 2
  status: 0
 success: True
       x: array([0.39999999, 0.60000001])

In [26]:
round(res['fun'], 5)

0.0

In [27]:
# x is from array shown above
Q = res['x'].round(5) 
Q

array([0.4, 0.6])

In [28]:
error(Q)

0.0

In [29]:
np.dot(S,Q) / (1+i)

10.0

In [30]:
S0

10.0

### Risk Neutral ('Martingale') Pricing 

In [31]:
phi

array([ 0.33333333, -0.15151515])

In [32]:
np.dot(M, phi)

array([5.00000000e+00, 2.22044605e-16])

In [33]:
np.dot(M0, phi)

1.8181818181818188

In [34]:
np.dot(C, P) / (1+i) # wrong measure OR wrong discount rate

2.2727272727272725

In [35]:
np.dot(C, Q) / (1+i) # martingale meas in combo with risk free rate

1.8181818181818181

### Aproximation by Ordinary Least Squares

In [36]:
reg = np.linalg.lstsq(M,C)[0]
reg

/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  """Entry point for launching an IPython kernel.


array([ 0.33333333, -0.15151515])

In [37]:
np.dot(M, reg).round(5)

array([5., 0.])

In [38]:
np.dot(M0, reg)

1.8181818181818188

### Option Pricing Based on Neural Nets ("Learning')

In [39]:
w = np.array((1., -1.))

In [40]:
l1 = np.dot(M, w)  
l1

array([ 9., -6.])

In [41]:
d = l1 - C
d

array([ 4., -6.])

In [42]:
#alpha = 0.01

In [43]:
w -= alpha * d
w

NameError: name 'alpha' is not defined

# I stopped at 1 hr 3 minuts